### PR TITLE
o/state,daemon: add snap-run-inhibit notice

### PR DESCRIPTION
**Pull request by ZeyadYasser**
This PR adds the `snap-run-inhibit` notice which is need for the "snap-run during refresh UX" flow described in [SD167](https://docs.google.com/document/d/1HJQWKzgaB3yPKwRUqlxYhgyaKG5IcJe8eZOT4YY1CI8/edit#heading=h.v228s8hnqd53).

This is mostly a port of `custom` notices implementation from pebble https://github.com/canonical/pebble/pull/296 adapted for the `snap-run-inhibit` notice.

This PR is a building block for https://github.com/snapcore/snapd/pull/13770.

### CODE REVIEW

**Review by olivercalder — CHANGES_REQUESTED** [review]
Thanks for your work on this. I'm rather skeptical of the approach of allowing the creation of `"snap-run-inhibit"` notice by posting to the notices endpoint from outside of snapd. This seems to me like something which should be determined internally by snapd. As it is now, any client/user would be able to create a `"snap-run-inhibit"` notice by simply posting to the notices API and accepting the polkit prompt, as far as I understand it.

It really feels to me like snapd should be responsible for deciding when the `"snap-run-inhibit"` notice should be issued. Maybe that means posting to another API endpoint, which then tells snapd to record the notice, but posting to the notices endpoint should record a notice without any complicated checks beyond validating the type and data format. I think for this notice type deeper checks would seem to be required, which feels beyond the scope of posting to the notices endpoint, in my opinion.

Perhaps I'm misunderstanding though. Is there a reason we really need to post to the notices endpoint for this use case?

File: daemon/api_notices.go [hunk 1]
```diff
@@ -31,13 +32,16 @@ import (
...
 var (
 	noticesCmd = &Command{
-		Path:       "/v2/notices",
-		GET:        getNotices,
-		ReadAccess: interfaceOpenAccess{Interface: "snap-refresh-observe"},
+		Path:        "/v2/notices",
+		GET:         getNotices,
+		POST:        postNotices,
+		ReadAccess:  interfaceOpenAccess{Interface: "snap-refresh-observe"},
+		WriteAccess: authenticatedAccess{Polkit: polkitActionManage},
```
[the remark this answers is not in the document]
Comment (ZeyadYasser) [reply daemon/api_notices.go:44]: > I'm a bit unsure as to whether polkit authentication makes sense here

Good point regarding polkit, The polkit prompting doesn't make sense in this use case, I will remove it.

> if I understand correctly, these notices are issued by the snap client when a snap is run, to inhibit refreshes?

No, They are issued by the snap client to announce that **it is being inhibited from running** due to an ongoing refresh, so that a desktop client (specifically `snapd-desktop-integration` snap) can listen for this notice so it can display notifications or refresh progress as described in [SD167](https://docs.google.com/document/d/1HJQWKzgaB3yPKwRUqlxYhgyaKG5IcJe8eZOT4YY1CI8/edit#heading=h.g0v0i714voiv).

> Here, it checks whether the connection has the "snap-refresh-observe" interface connected, but the notice endpoint itself is on the usual snapd.socket, not snapd-snap.socket, and doesn't require a particular interface?

Thanks for pointing this out, It does look weird without context, It warrants a comment explaining it.
This marker interface check is just to be able do UX fallback:
- if no one is listening for the notices (i.e. `snap-refresh-observe` marker interface is not connected), snap run sends the notifications itself (https://github.com/snapcore/snapd/commit/63634cc73400e13c0dc6d286d8f62f2f4bae2246#diff-2040a7ab4147eafb94c9b58c40202df922bd298e0ec3d152e322cef32f051db5R115-R131).

We could send the notice regardless of the interface check (which actually might be better for readability and consistency) that's why it is not relevant to the API implementation, but I just chose the route with less API calls.

File: daemon/api_notices.go [hunk 2]
```diff
@@ -218,6 +230,53 @@ func allowedNoticeTypesForInterface(iface string) []state.NoticeType {
 	return types
 }
 
+func postNotices(c *Command, r *http.Request, user *auth.UserState) Response {
```
[the remark this answers is not in the document]
Comment (ZeyadYasser) [reply daemon/api_notices.go:233]: > As far as I understand, this notice type is something produced from within snapd

Not really, `snap-run-inhibit` indicate that `snap run` couldn't start because of an ongoing refresh. snapd have no way of telling when `snap run` run is called.

Thank you for pointing this out, it means that I should document this better.

File: daemon/api_notices.go [hunk 3]
```diff
@@ -218,6 +230,53 @@ func allowedNoticeTypesForInterface(iface string) []state.NoticeType {
...
+	}
+	decoder := json.NewDecoder(r.Body)
+	if err := decoder.Decode(&payload); err != nil {
+		return BadRequest("cannot decode request body: %v", err)
+	}
+
+	if payload.Action != "add" {
+		return BadRequest("invalid action %q", payload.Action)
+	}
+	if payload.Type != "snap-run-inhibit" {
```
[the remark this answers is not in the document]
Comment (ZeyadYasser) [reply daemon/api_notices.go:253]: Actually, we were going to use `custom` notices for this, but the fact we had to stuff the notice type (snap run is inhibited) and key (inhibited instance name) in the `custom` notice key (e.g. `key="snap-run-inhibited/firefox"`) meant that a listener (i.e. `snapd-desktop-integration`) won't have a way to filter for `snap-run-inhibited` events since we don't have pattern based matching for notices (yet). Also, we wanted to be less flexible with custom notices since it doesn't make sense for snapd, unlike pebble. 

Basically, it is custom notices, but with a type field.

**Review by olivercalder — COMMENTED** [review]
Some thoughts after discussing with @ZeyadYasser:

- I don't think we want to use `authenticatedAccess` on any http method of the `/v2/notices` endpoint
  - Notices should be open to anyone (up to user and interface requirements) to read, and recording notices should be similarly open
  - To do otherwise would preclude future custom notice types which applications should be able to issue without authentication
  - Thus, if we want to add a `POST /v2/notices` endpoint, it should be `openAccess` or `interfaceOpenAccess`
- Since `"snap-run-inhibit"` notices are supposed to tell something which is reliably true about the fact that a snap cannot currently be run, a non-privileged application (snap or not) should not be able to erroneously or maliciously create a `"snap-run-inhibit"` notice by simply posting to the open notices API
  - I believe this would be less verifiable than the current system of `snap run` sending a dbus message to `snapd-desktop-integration`, but even if not, I think we can do better at ensuring that the `"snap-run-inhibit"` notice is created because of `snap run`, not a third party
  - I think snapd should verify this as well as possible, and then add the `"snap-run-inhibit"` notice directly/internally
- I would propose creating a dedicated snapd API endpoint for the `snap` client to tell snapd that `snap run` is inhibited, and then snapd should be responsible for recording the `"snap-run-inhibit"` notice
  - This endpoint could be, for example, `/v2/snaps/{name}/inhibit`
    - Here, the snap name would be included in the path, and would not need to be parsed from the request body
  - Such an endpoint should be `authenticatedAccess`, and snapd should verify that the request comes from the `snap` client, not a third party
    - No polkit authentication, probably should use a macaroon, but I'm not sure how those work
      - Is `authenticatedAccess` with a macaroon for authentication possible, and if so, what makes it meaningfully more secure than `openAccess`?
  - If possible, it would be great if snapd could verify that there is actually a refresh occurring for the snap in question, and if so, then it creates the `"snap-run-inhibit"` notice on behalf of the `snap` client
  - This way, `POST /v2/notices` would still be available in the future if we want to add more permissive notice types that users/applications can create at will
  - In the meantime, we can remove `POST /v2/notices`, `client/notices.go`, etc.

I'd very much appreciate hearing what @pedronis thinks of this approach, as it does contradict some of the existing RAA spec.

File: daemon/api_notices.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (pedronis) [reply daemon/api_notices.go]: I don't think we want authentication here as such

**Comment by pedronis** [comment]
@olivercalder @ZeyadYasser 

the idea was always to allow to create  this notice only at least if the requester /proc/PID/exe points to a snapd "snap" command, there might be more sophisticated possible checks, but that would be the minimum

File: daemon/api_notices.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ZeyadYasser) [reply daemon/api_notices.go]: Updated, Thank you!

**Comment by ZeyadYasser** [comment]
@pedronis @olivercalder  Added the check for requester `/proc/PID/exe` as suggested here https://github.com/snapcore/snapd/pull/13769#issuecomment-2031758416

**Comment by ZeyadYasser** [comment]
Rebased to resolve conflicts as they were preventing tests from running.

File: daemon/api_notices.go [hunk 4]
```diff
@@ -16,28 +16,34 @@ package daemon
...
 var (
 	noticesCmd = &Command{
-		Path:       "/v2/notices",
-		GET:        getNotices,
-		ReadAccess: interfaceOpenAccess{Interfaces: []string{"snap-refresh-observe"}},
+		Path:        "/v2/notices",
+		GET:         getNotices,
+		POST:        postNotices,
+		ReadAccess:  interfaceOpenAccess{Interfaces: []string{"snap-refresh-observe"}},
+		WriteAccess: openAccess{},
```
Comment (olivercalder) [inline daemon/api_notices.go:46]: I think perhaps a different type of access, something like `snapClientAccess{}` might be appropriate here? Again, for the time being, we don't want to allow anyone outside of snapd/the snap client to be able to create notices. Your solution of checking `/proc/<pid>/exe` works, but the endpoint being `openAccess` suggests to me that anyone should be able to create notices of any type, neither of which are true.

Something like `snapClientAccess` could implement the check you added below while maintaining that this behaves differently from the way `openAccess` works in other places, (afaik) without restrictions about the caller (beyond it not being a snap).

File: daemon/api_notices.go [hunk 5]
```diff
@@ -232,6 +246,104 @@ func allowedNoticeTypesForInterface(iface string) []state.NoticeType {
...
+
+	noticeId, err := st.AddNotice(&requestUID, state.SnapRunInhibitNotice, payload.Key, nil)
+	if err != nil {
+		return InternalError("%v", err)
+	}
+
+	return SyncResponse(addedNotice{ID: noticeId})
+}
+
+func requestFromSnapCmd(st *state.State, r *http.Request) (bool, error) {
```
Comment (olivercalder) [inline daemon/api_notices.go:304]: This is a good check, thanks! I think it would be great as a general check for a new access checker like `snapClientAccess{}`.

File: daemon/api_notices_test.go [hunk 6]
```diff
@@ -275,20 +284,21 @@ func (s *noticesSuite) TestNoticesFilterTypesForSnap(c *C) {
...
 	addNotice(c, st, nil, state.WarningNotice, "danger", nil)
+	addNotice(c, st, nil, state.SnapRunInhibitNotice, "snap-name", nil)
 	st.Unlock()
 
 	// Check that a snap request with types filter allows access to
 	// snaps with required interfaces only.
 
-	// snap-refresh-observe interface allows accessing change-update notices
-	req, err := http.NewRequest("GET", "/v2/notices?types=change-update,refresh-inhibit", nil)
+	// snap-refresh-observe interface allows accessing change-update, refresh-inhibir and snap-run-inhibit notices
```
Comment (olivercalder) [inline daemon/api_notices_test.go:293]: s/refresh-inhibir/refresh-inhibit/

**Review by olivercalder — COMMENTED** [review]
Looks really good, thanks! I think given the very special handling of "only run inhibit notice" and "only snap client", we could use a more specialized access checker than `openAccess`. What would you think of this @pedronis ?

File: daemon/api_notices_test.go (hunk 6 above)
Comment (ZeyadYasser) [reply daemon/api_notices_test.go]: Good catch, thanks! updated

File: daemon/api_notices.go (hunk 4 above)
Comment (olivercalder) [reply daemon/api_notices.go:46]: That's a good point. Furthermore, in the future, we may wish to allow snaps to create notices, in which case it would become `interfaceOpenAccess`, which is definitely not compatible with a `snapClientAccess`. Using `openAccess` as you suggest is analogous to the way `interfaceOpenAccess` grants access only to particular notice types depending on the connected interface, so there is precedent as well.

So +1 from me for your current approach. Thanks!

**Review by olivercalder — APPROVED** [review]
Looks good to me, thanks for the discussion.

File: daemon/api_notices.go (hunk 4 above)
Comment (Meulengracht) [reply daemon/api_notices.go:46]: How is this endpoint currently protected in pebble?

File: daemon/api_notices.go [hunk 7]
```diff
@@ -232,6 +246,104 @@ func allowedNoticeTypesForInterface(iface string) []state.NoticeType {
...
+	var payload struct {
+		Action string `json:"action"`
+		Type   string `json:"type"`
+		Key    string `json:"key"`
+		// NOTE: Data and RepeatAfter fields are not needed for snap-run-inhibit notices.
+	}
+	decoder := json.NewDecoder(r.Body)
+	if err := decoder.Decode(&payload); err != nil {
+		return BadRequest("cannot decode request body: %v", err)
+	}
```
Comment (Meulengracht) [inline daemon/api_notices.go:264]: This is definitely not a nice solution - this distracts from the purpose of the postNotices handler, and creates a lot of noise. Please move `var payload struct` out of the function, give it a more meaningful name and a docstring explaining why it's needed/purpose, and then create a function called `decode<>AndValidate`, where you move the validation logic to as well. Having a separate `decode` and `validate` function here is acceptable as well.

File: daemon/api_notices.go [hunk 8]
```diff
@@ -232,6 +246,104 @@ func allowedNoticeTypesForInterface(iface string) []state.NoticeType {
...
+	st.Lock()
+	defer st.Unlock()
+
+	fromSnapCmd, err := requestFromSnapCmd(st, r)
+	if err != nil {
+		return InternalError("cannot check request source")
+	}
+	if !fromSnapCmd {
+		return Forbidden("only snap command can record notices")
+	}
```
Comment (Meulengracht) [inline daemon/api_notices.go:286]: Since they aren't used further down, we can scope the variables here
```suggestion
	if fromSnapCmd, err := requestFromSnapCmd(st, r); err != nil {
		return InternalError("cannot check request source")
	} else  if !fromSnapCmd {
		return Forbidden("only snap command can record notices")
	}
```

File: daemon/api_notices.go [hunk 9]
```diff
@@ -232,6 +246,104 @@ func allowedNoticeTypesForInterface(iface string) []state.NoticeType {
...
+
+	exists, err := snapInstanceExists(st, payload.Key)
+	if err != nil {
+		return InternalError("cannot check snap in state: %v", err)
+	}
+	if !exists {
+		return BadRequest("snap %q does not exist", payload.Key)
+	}
```
Comment (Meulengracht) [inline daemon/api_notices.go:294]: Same here
```suggestion
	if exists, err := snapInstanceExists(st, payload.Key); err != nil {
		return InternalError("cannot check snap in state: %v", err)
	} else if !exists {
		return BadRequest("snap %q does not exist", payload.Key)
	}
```

File: daemon/api_notices.go [hunk 10]
```diff
@@ -47,6 +53,14 @@ var (
 	}
 )
 
+const (
+	maxNoticeKeyLength = 256
+)
+
+type addedNotice struct {
```
Comment (Meulengracht) [inline daemon/api_notices.go:60]: This needs a docstring

File: daemon/api_notices.go [hunk 11]
```diff
@@ -47,6 +53,14 @@ var (
 	}
 )
 
+const (
+	maxNoticeKeyLength = 256
```
Comment (Meulengracht) [inline daemon/api_notices.go:57]: Should this be a more common place for notices? I'm not sure I feel like this would be the correct place to have it. Also do we have some more notice-key validation logic somewhere it could fit?

**Review by Meulengracht — COMMENTED** [review]
I did a first pass

File: daemon/api_notices.go (hunk 7 above)
Comment (pedronis) [reply daemon/api_notices.go:264]: this code is like this in pebble:

https://github.com/canonical/pebble/blob/master/internals/daemon/api_notices.go#L200

I have less of a strong opinion on this, as the struct is not used across many endpoints. Anyway if we change this we would need to backport the change to pebble

Comment (pedronis) [reply daemon/api_notices.go:264]: this style is used in a few places in daemon fwiw

File: daemon/api_notices.go [hunk 12]
```diff
@@ -232,6 +246,104 @@ func allowedNoticeTypesForInterface(iface string) []state.NoticeType {
...
+	decoder := json.NewDecoder(r.Body)
+	if err := decoder.Decode(&payload); err != nil {
+		return BadRequest("cannot decode request body: %v", err)
+	}
+
+	if payload.Action != "add" {
+		return BadRequest("invalid action %q", payload.Action)
+	}
+	if payload.Type != "snap-run-inhibit" {
+		return BadRequest(`invalid type %q (can only add "snap-run-inhibit" notices)`, payload.Type)
```
Comment (pedronis) [inline daemon/api_notices.go:270]: this error is probably confusing, we can only decide what to do with the type once we have check the command things are coming from below

File: daemon/api_notices.go (hunk 5 above)
Comment (pedronis) [reply daemon/api_notices.go:304]: I think isRequestFrom would be a bit a clearer name for this, also it migh belong to access.go

File: daemon/request.go [hunk 13]
```diff
@@ -15,9 +15,12 @@
 package daemon
 
 import (
+	"os"
 	"time"
 )
 
+var osReadlink = os.Readlink
```
Comment (pedronis) [inline daemon/request.go:22]: unless it is used in multiple places this should go in the file that needs it

**Review by pedronis — COMMENTED** [review]
did a first pass

File: daemon/api_notices.go (hunk 4 above)
Comment (ZeyadYasser) [reply daemon/api_notices.go:46]: @Meulengracht  It is not protected, it uses an access checker that similar to `openAccess`.

File: daemon/api_notices.go (hunk 7 above)
Comment (ZeyadYasser) [reply daemon/api_notices.go:264]: I believe for snapd it makes more sense to have the `decode<>AndValidate` approach because we might have other notices (e.g. `custom` notices someday), but in I don't believe it make sense for pebble since the postNotices endpoint whole purpose is to support `custom` notices, it was built for specific purpose.

I understand the appeal of using a pattern that works for both snapd and pebble, but I think input from @benhoyt would be appreciated.

File: daemon/access.go [hunk 14]
```diff
@@ -255,3 +263,50 @@ func (ac interfaceAuthenticatedAccess) CheckAccess(d *Daemon, r *http.Request, u
...
+	}
+
+	// Check if re-exec in snapd
+	var snapst snapstate.SnapState
+	err = snapstate.Get(st, "snapd", &snapst)
+	if err != nil && !errors.Is(err, state.ErrNoState) {
+		return false, err
+	}
+	if err == nil {
+		snapdPath := fmt.Sprintf("snapd/%s/usr/bin/snap", snapst.CurrentSideInfo().Revision)
```
Comment (bboozzoo) [inline daemon/access.go:293]: does it have to be current, or would any installed revision be acceptable? i.e. what if snapd is refreshing just now.

File: daemon/api_notices.go (hunk 7 above)
Comment (ZeyadYasser) [reply daemon/api_notices.go]: Updated, thanks!

File: daemon/api_notices.go (hunk 8 above)
Comment (ZeyadYasser) [reply daemon/api_notices.go]: updated

File: daemon/api_notices.go (hunk 10 above)
Comment (ZeyadYasser) [reply daemon/api_notices.go:55]: Good point, thanks! updated

File: daemon/api_notices.go (hunk 11 above)
Comment (ZeyadYasser) [reply daemon/api_notices.go]: I moved the validation to `state/notices.go` and shared it with the daemon, I like this better, Thanks!

File: daemon/api_notices.go (hunk 12 above)
Comment (ZeyadYasser) [reply daemon/api_notices.go]: Good point, Thanks! updated

File: overlord/state/notices.go [hunk 15]
```diff
@@ -279,15 +285,19 @@ func (s *State) AddNotice(userID *uint32, noticeType NoticeType, key string, opt
 	return notice.id, nil
 }
 
-func validateNotice(noticeType NoticeType, key string, options *AddNoticeOptions) error {
+// ValidateNotice validates notice type and key before adding.
+func ValidateNotice(noticeType NoticeType, key string, options *AddNoticeOptions) error {
 	if !noticeType.Valid() {
-		return fmt.Errorf("internal error: attempted to add notice with invalid type %q", noticeType)
+		return fmt.Errorf("attempted to add notice with invalid type %q", noticeType)
```
Comment (pedronis) [inline overlord/state/notices.go:291]: I think all the "attempted to" here should be turned into "cannot"  if we remove the "internal error" prefix

**Review by pedronis — COMMENTED** [review]
couple small comments

File: overlord/state/notices.go (hunk 15 above)
Comment (ZeyadYasser) [reply overlord/state/notices.go]: updated, thanks!

File: daemon/access.go (hunk 14 above)
Comment (ZeyadYasser) [reply daemon/access.go]: updated to use wildcard matching, great catch, thanks!

**Review by Meulengracht — APPROVED** [review]
Thank you

File: daemon/api_notices.go [hunk 16]
```diff
@@ -232,6 +242,73 @@ func allowedNoticeTypesForInterface(iface string) []state.NoticeType {
...
+}
+
+func (inst *noticeInstruction) validateSnapRunInhibitNotice(st *state.State, r *http.Request) *apiError {
+	if fromSnapCmd, err := isRequestFromSnapCmd(r); err != nil {
+		return InternalError("cannot check request source: %v", err)
+	} else if !fromSnapCmd {
+		return Forbidden("only snap command can record notices")
+	}
+
+	if exists, err := snapInstanceExists(st, inst.Key); err != nil {
```
[the remark this answers is not in the document]
Comment (bboozzoo) [reply daemon/api_notices.go:303]: just FYI, feel free to push this change in a followup
